synapse/handlers/__init__.py

```diff
@@ -54,7 +55,12 @@ def __init__(self, hs):
         self.directory_handler = DirectoryHandler(hs)
         self.typing_notification_handler = TypingNotificationHandler(hs)
         self.admin_handler = AdminHandler(hs)
+        asapi = ApplicationServiceApi(hs)
         self.appservice_handler = ApplicationServicesHandler(
-            hs, ApplicationServiceApi(hs)
+            hs, asapi, AppServiceScheduler(
+                clock=hs.get_clock(),
+                store=hs.get_datastore(),
+                as_api=asapi
+            )
```

Comment:
I dislike the idea of putting every single DI component in ``HomeServer`` like this because it exposes them as effectively global variables. No other component needs to (or should need to) use ``AppServiceScheduler`` except ``ApplicationServicesHandler``. Therefore, I strongly think that we shouldn't be exposing it in ``HomeServer``. We should be *selectively* exposing things as global, rather than making it global by default simply so you can DI it in my opinion. If you still feel like it should be done for consistency sake though, then fair enough.